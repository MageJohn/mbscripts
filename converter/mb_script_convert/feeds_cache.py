import atexit
import pickle
from os import environ
from pathlib import Path
from typing import Any

XDG_CACHE_HOME = Path(environ.get("XDG_CACHE_HOME", Path.home() / ".cache"))
CACHE_FILE = XDG_CACHE_HOME / "mb-scripts" / "cache.pickle"


class Cache(dict[str, Any]):
    def __init__(self, cache_file: str | Path):
        self.modified: bool = False
        self.cache_file = Path(cache_file)
        return super().__init__()

    def __setitem__(self, key: str, value: Any, /) -> None:
        self.modified = True
        return super().__setitem__(key, value)

    def load_cache(self):
        if self.cache_file.exists():
            with self.cache_file.open("rb") as f:
                cache: Cache = pickle.load(f)
                self.update(cache)

    def save_cache(self):
        if self.modified:
            self.cache_file.parent.mkdir(parents=True, exist_ok=True)
            with self.cache_file.open("wb") as f:
                pickle.dump(self, f, pickle.HIGHEST_PROTOCOL)


cache = Cache(CACHE_FILE)
cache.load_cache()
atexit.register(cache.save_cache)
