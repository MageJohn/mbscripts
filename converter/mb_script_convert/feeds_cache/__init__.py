import atexit
from os import environ
from pathlib import Path

from .cache import Cache

XDG_CACHE_HOME = Path(environ.get("XDG_CACHE_HOME", Path.home() / ".cache"))
CACHE_FILE = XDG_CACHE_HOME / "mb-scripts" / "cache.pickle"

cache = Cache(CACHE_FILE)
cache.load_cache()
atexit.register(cache.save_cache)
