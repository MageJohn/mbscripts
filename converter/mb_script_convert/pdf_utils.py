from __future__ import annotations

import re
from math import isclose
from typing import TYPE_CHECKING

from py_pdf_parser.common import BoundingBox

if TYPE_CHECKING:
    from collections.abc import Callable

    from py_pdf_parser.filtering import PDFElement

PAGE_X_CENTER = 500
PAGE_Y_CENTER = 740


def is_centered(element: PDFElement) -> bool:
    page = element.document.get_page(element.page_number)
    page_middle = page.width / 2

    bbox = element.bounding_box
    el_middle = (bbox.x0 + bbox.x1) / 2

    return isclose(page_middle, el_middle, rel_tol=0.015)


def is_in_top_right(element: PDFElement) -> bool:
    """If it is the top right of the page, could be a page number."""
    bb = element.bounding_box
    return bb.x0 >= PAGE_X_CENTER and bb.y0 >= PAGE_Y_CENTER


def is_off_page(element: PDFElement) -> bool:
    page = element.document.get_page(element.page_number)
    page_bb = BoundingBox(0, page.width, 0, page.height)
    return not element.partially_within(page_bb)


def by_indent(indent: float | int) -> Callable[[PDFElement], bool]:
    def predicate(element: PDFElement) -> bool:
        return isclose(element.bounding_box.x0, indent, abs_tol=10)

    return predicate


def by_min_indent(indent: float | int) -> Callable[[PDFElement], bool]:
    def predicate(element: PDFElement) -> bool:
        return element.bounding_box.x0 > indent

    return predicate


def clean_text(element: PDFElement) -> str:
    return join_lines(element.text())


LINE_BREAK = re.compile(r" *\n *")
HYPHEN_LINE_BREAK = re.compile(r"-\n")


def join_lines(string: str) -> str:
    joined_hyphens = HYPHEN_LINE_BREAK.sub("-", string)
    return LINE_BREAK.sub(" ", joined_hyphens)


class NotTaggedError(Exception):
    pass
