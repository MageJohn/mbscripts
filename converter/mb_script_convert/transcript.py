import logging
from dataclasses import dataclass, fields
from datetime import datetime
from typing import Literal, Optional, Self

logger = logging.getLogger(__name__)

type Tag = Literal["direction", "character", "parenthetical", "dialogue", "end"]
type Content = list[tuple[Tag, str]]


type HugoFrontmatter = dict[str, str | HugoFrontmatter]


@dataclass
class Metadata:
    episode_title: Optional[str] = None
    series: Optional[str] = None
    season: Optional[int] = None
    season_episode_number: Optional[int] = None
    date_published: Optional[datetime] = None
    cover_url: Optional[str] = None

    def merge(self, other: Self):
        for field in fields(self):
            if getattr(self, field.name) is None:
                setattr(self, field.name, getattr(other, field.name))


class Transcript:
    """
    Represents a transcript in a format agnostic way.
    Used as an intermediate step when converting.
    """

    content: Content
    metadata: Metadata

    def __init__(
        self, metadata: Metadata | None = None, content: Content | None = None
    ):
        self.metadata = Metadata() if metadata is None else metadata
        self.content = [] if content is None else content

    def add_content(self, tag: Tag, content: str):
        self.content.append((tag, content))
