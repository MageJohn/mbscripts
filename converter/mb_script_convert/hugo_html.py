import logging
import tomllib
from contextlib import nullcontext
from io import BytesIO
from pathlib import Path
from typing import IO, Union

import lxml.html.builder as b
import tomli_w
from adaptix import P, Retort, as_sentinel, name_mapping
from lxml import html

from .transcript import Metadata, Transcript

logger = logging.getLogger(__name__)


retort = Retort(
    recipe=[
        name_mapping(
            Metadata,
            map=[
                ("episode_title", "title"),
                ("date_published", "date"),
                (".*", ["params", ...]),
            ],
            omit_default=True,
        ),
        as_sentinel(P[Metadata][Union][None]),
    ],
    strict_coercion=False,
)


def dump(transcript: Transcript, file_or_path: str | Path | IO[bytes]):
    if isinstance(file_or_path, (str, Path)):
        path = Path(file_or_path)
        path.parent.mkdir(parents=True, exist_ok=True)
        file = path.open("wb+")
    else:
        file = nullcontext(file_or_path)
    with file as fh:
        fh.write(b"+++\n")
        tomli_w.dump(retort.dump(transcript.metadata), fh)
        fh.write(b"+++\n\n")

        for el in transcript.content:
            html_el = b.P(b.CLASS(el[0]), el[1])
            html_el_str = html.tostring(html_el, pretty_print=True, encoding="utf-8")
            assert isinstance(html_el_str, bytes)  # for type hinting
            fh.write(html_el_str)


def dumps(transcript: Transcript) -> str:
    out = BytesIO()
    dump(transcript, out)
    return out.getvalue().decode("utf-8")


def load_metadata(file_or_path: str | Path | IO[str], transcript: Transcript):
    if isinstance(file_or_path, (str, Path)):
        path = Path(file_or_path)
        assert path.exists(), "Invalid path passed"
        file = path.open("r", encoding="utf-8")
    else:
        file = nullcontext(file_or_path)
    with file as fh:
        if fh.readline() != "+++\n":
            logger.warning("Existing output file is not a valid Hugo document")
        frontmatter_str = ""
        while (line := fh.readline()) != "+++\n":
            frontmatter_str += line

    frontmatter = tomllib.loads(frontmatter_str)
    transcript.metadata = retort.load(frontmatter, Metadata)
