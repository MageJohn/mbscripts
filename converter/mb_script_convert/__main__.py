import argparse
import logging
import sys

from .main import main

logger = logging.getLogger(__name__)

MB_RSS_URL = "https://rss.art19.com/midnight-burger"

parser = argparse.ArgumentParser()
parser.add_argument("IN")
parser.add_argument("-o", "--output", default=sys.stdout.buffer)
parser.add_argument(
    "--episode-title", help="Override the episode title as parsed from the PDF"
)
parser.add_argument(
    "--skip-scraping",
    action="store_true",
    help="Skip scraping extra metadata from the RSS feed",
)
parser.add_argument(
    "--rss-url",
    default=MB_RSS_URL,
    help="Override the URL of the Midnight Burger RSS feed",
)
parser.add_argument(
    "-O", "--overwrite", action="store_true", help="Allow overwriting the output"
)
parser.add_argument("--debug", action="store_true")

if __name__ == "__main__":
    args = parser.parse_args()
    try:
        main(
            args.IN,
            args.output,
            args.episode_title,
            args.skip_scraping,
            args.rss_url,
            args.overwrite,
            args.debug,
        )
    except RuntimeError as e:
        logger.error(*e.args)
