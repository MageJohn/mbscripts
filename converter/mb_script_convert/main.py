import logging
from pathlib import Path

from mb_script_convert.scrape_rss import scrape_episode_metadata

from . import normalisations
from .hugo_html import dump, load_metadata
from .import_midnight_burger import import_transcript

logger = logging.getLogger(__name__)


def main(
    in_file_or_dir: str,
    out_file_or_dir: str,
    episode_title: str | None,
    skip_scraping: bool,
    rss_url: str | None,
    overwrite: bool,
    debug: bool,
):
    logging.basicConfig(level=logging.INFO, format="{levelname}: {message}", style="{")

    in_path = Path(in_file_or_dir)
    out_path = Path(out_file_or_dir)

    if in_path.is_dir():
        if not out_path.is_dir():
            raise RuntimeError(
                "when input is a directory, output must also be a directory"
            )
        for pdf_path in in_path.glob("**/*.pdf"):
            output_file_path = make_output_path(pdf_path, in_path, out_path)
            if not overwrite and output_file_path.exists():
                logger.info(f"Skipping: {pdf_path} -> {output_file_path}")
                continue
            convert_transcript(
                pdf_path,
                output_file_path,
                episode_title,
                skip_scraping,
                rss_url,
                debug,
            )
    elif in_path.is_file():
        if not overwrite and out_path.exists():
            logger.info(f"Skipping: {in_path} -> {out_path}")
        else:
            convert_transcript(
                in_path, out_path, episode_title, skip_scraping, rss_url, debug
            )
    else:
        raise RuntimeError(f"not a valid file or directory: {in_file_or_dir}")


def make_output_path(input: Path, input_base: Path, output_base: Path) -> Path:
    """Make an output path for a converted PDF

    Will recreate the relative path from `input_base` to `input` in
    `output_base`, with `.html` replacing `.pdf`. However, if
    `name/index.html` exists in the output_base path, then return that
    `index.html` file to ensure proper overwriting behaviour.
    Arguments:
    input -- path to input PDF file; must be a subpath of `input_base`
    input_base -- parent path of `input`; the path below this will be recreated in `output_path`
    output_base -- the target directory where the output path will be
    """
    rel_input = input.relative_to(input_base)
    output_html = output_base / rel_input.with_suffix(".html")
    output_indexhtml = output_base / rel_input.with_name(input.stem) / "index.html"

    if output_indexhtml.exists():
        return output_indexhtml
    else:
        return output_html


def convert_transcript(
    input: Path,
    output: Path,
    episode_title: str | None,
    skip_scraping: bool,
    rss_url: str | None,
    debug: bool,
):
    logger.info(f"Converting: {input} -> {output}")
    doc = import_transcript(str(input), debug)
    normalisations.run_all(doc)
    if output.exists():
        load_metadata(output, doc)
    if episode_title:
        doc.metadata.episode_title = episode_title
    if not skip_scraping:
        scrape_episode_metadata(doc, rss_url)
    dump(doc, output)
