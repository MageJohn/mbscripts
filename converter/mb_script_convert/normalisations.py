"""Functions which normalise the text content of a Transcript."""

from mb_script_convert.transcript import Transcript


def run_all(transcript: Transcript):
    title_period(transcript)
    series_colon(transcript)


def title_period(transcript: Transcript):
    title = transcript.metadata.episode_title
    if title and title[-1] == ".":
        transcript.metadata.episode_title = title[:-1]


def series_colon(transcript: Transcript):
    series = transcript.metadata.series
    if series and series[-1] == ":":
        transcript.metadata.series = series[:-1]
