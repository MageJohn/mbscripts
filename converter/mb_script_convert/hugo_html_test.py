from datetime import datetime
from pathlib import Path

import pytest

from mb_script_convert.hugo_html import dump, dumps, load_metadata
from mb_script_convert.transcript import Metadata, Transcript


@pytest.fixture
def transcript():
    return Transcript(
        metadata=Metadata(
            episode_title="Panopticon",
            season=4,
            date_published=datetime.fromisoformat("2019-10-17"),
        ),
        content=[
            ("direction", "TAPE CLICKS ON"),
            ("character", "PETER"),
            ("parenthetical", "(pleasantly)"),
            ("dialogue", "Is everything alright, Martin?"),
        ],
    )


def test_dumps(transcript, snapshot):
    assert dumps(transcript) == snapshot


def test_dump(transcript, snapshot, tmp_path: Path):
    out_path = tmp_path / "out_Path.html"
    out_str = (tmp_path / "out_str.html").as_posix()
    out_fh_path = tmp_path / "out_fh.html"

    dump(transcript, out_path)
    dump(transcript, out_str)
    with open(out_fh_path, "wb") as out_fh:
        dump(transcript, out_fh)

    with open(out_path, "r") as fh:
        assert fh.read() == snapshot
    with open(out_str, "r") as fh:
        assert fh.read() == snapshot
    with open(out_fh_path, "r") as fh:
        assert fh.read() == snapshot


def test_load_metadata(transcript, tmp_path: Path):
    existing_transcript = tmp_path / "test.html"
    dump(transcript, existing_transcript)

    new_transcript = Transcript()
    load_metadata(existing_transcript, new_transcript)

    assert new_transcript.metadata == transcript.metadata
