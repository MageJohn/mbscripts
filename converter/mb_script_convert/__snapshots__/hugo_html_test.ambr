# serializer version: 1
# name: test_dump
  '''
  +++
  title = "Panopticon"
  date = "2019-10-17T00:00:00"
  
  [params]
  season = 4
  +++
  
  <p class="direction">TAPE CLICKS ON</p>
  <p class="character">PETER</p>
  <p class="parenthetical">(pleasantly)</p>
  <p class="dialogue">Is everything alright, Martin?</p>
  
  '''
# ---
# name: test_dump.1
  '''
  +++
  title = "Panopticon"
  date = "2019-10-17T00:00:00"
  
  [params]
  season = 4
  +++
  
  <p class="direction">TAPE CLICKS ON</p>
  <p class="character">PETER</p>
  <p class="parenthetical">(pleasantly)</p>
  <p class="dialogue">Is everything alright, Martin?</p>
  
  '''
# ---
# name: test_dump.2
  '''
  +++
  title = "Panopticon"
  date = "2019-10-17T00:00:00"
  
  [params]
  season = 4
  +++
  
  <p class="direction">TAPE CLICKS ON</p>
  <p class="character">PETER</p>
  <p class="parenthetical">(pleasantly)</p>
  <p class="dialogue">Is everything alright, Martin?</p>
  
  '''
# ---
# name: test_dumps
  '''
  +++
  title = "Panopticon"
  date = "2019-10-17T00:00:00"
  
  [params]
  season = 4
  +++
  
  <p class="direction">TAPE CLICKS ON</p>
  <p class="character">PETER</p>
  <p class="parenthetical">(pleasantly)</p>
  <p class="dialogue">Is everything alright, Martin?</p>
  
  '''
# ---
